core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -1031,36 +1019,30 @@ private object RTS {
     }
 
     @tailrec
-    private final def join0(cb: Callback[E, A]): Async[E, A] = {
+    private final def observe0(cb: Callback[Nothing, ExitResult[E, A]]): Async[Nothing, ExitResult[E, A]] = {
       val oldStatus = status.get
 
       oldStatus match {
-        case s @ Executing(_, _, _, _, _) =>
-          val newStatus = s.copy(joiners = cb :: s.joiners)
+        case s @ Executing(_, _, _) =>
+          val newStatus = s.copy(observers = cb :: s.observers)
 
-          if (!status.compareAndSet(oldStatus, newStatus)) join0(cb)
-          else Async.later[E, A]
+          if (!status.compareAndSet(oldStatus, newStatus)) observe0(cb)
+          else Async.later[Nothing, ExitResult[E, A]]
 
-        case s @ AsyncRegion(_, _, _, _, _, _, _, _) =>
-          val newStatus = s.copy(joiners = cb :: s.joiners)
+        case s @ AsyncRegion(_, _, _, _, _, _) =>
+          val newStatus = s.copy(observers = cb :: s.observers)
 
-          if (!status.compareAndSet(oldStatus, newStatus)) join0(cb)
-          else Async.later[E, A]
+          if (!status.compareAndSet(oldStatus, newStatus)) observe0(cb)
+          else Async.later[Nothing, ExitResult[E, A]]
 
-        case Done(v) => Async.now(v)
+        case Done(v) => Async.now(ExitResult.Completed(v))
       }
     }
 
-    private final def purgeJoinersKillers(v: ExitResult[E, A],
-                                          exitHandlers: List[ExitResult[E, A] => IO[Nothing, Unit]],
-                                          joiners: List[Callback[E, A]],
-                                          killers: List[Callback[E, Unit]]): Unit = {
+    private final def purgeObservers(v: ExitResult[E, A], observers: List[Callback[Nothing, ExitResult[E, A]]]): Unit =
       // To preserve fair scheduling, we submit all resumptions on the thread
       // pool in (rough) order of their submission.
-      killers.reverse.foreach(k => rts.submit(k(SuccessUnit)))
-      joiners.foreach(k => rts.submit(k(v)))
-      exitHandlers.foreach(k => rts.unsafeRunAsync(k(v))((_: ExitResult[Nothing, Unit]) => ()))
-    }
+      observers.reverse.foreach(k => rts.submit(k(ExitResult.Completed(v))))
```

Comment:
This is really nice now, very clear ordering guarantees. You could now remove the `(rough)` in the Scaladoc if you wanted.